lib/api/apiUtils/authorization/permissionChecks.js

```diff
@@ -369,6 +369,54 @@ function isBucketAuthorized(bucket, requestTypes, canonicalID, authInfo, iamAuth
     return Object.keys(results).every(key => results[key] === true);
 }
 
+function evaluateBucketPolicyWithIAM(bucket, requestTypes, canonicalID, authInfo, iamAuthzResults, log, request) {
```

Comment:
In a better implementation we will want to avoid code duplication, feel free to refactor the whole file